### PR TITLE
Tracker for production-readiness approvals

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

**Comment by johnbelamaric** [comment]
I think this will be a conflict-magnet. Can we create a separate yaml per KEP, say `<kep-number>.yaml`?

**Comment by wojtek-t** [comment]
> I think this will be a conflict-magnet. Can we create a separate yaml per KEP, say <kep-number>.yaml?

I'm afraid we may end-up with gazillions files in the directory.
OTOH, they are easily discoverable my number, so actually that might be fine.
We may consider splitting by sig into directories at the later phase if needed too...

@johnbelamaric Assuming I will split that into per-kep file, does that look reasonable to you?

**Comment by johnbelamaric** [comment]
Yes. Go ahead and split them into SIG-based directories, and kep-per-file, by number.

Then, we need to update the template to tell people how to do it, and update kepval to check for it. You could even have kepctl create the file for a kep though that may be overkill. 

Thanks!

File: cmd/kepval/main_test.go
Comment (wojtek-t) [inline cmd/kepval/main_test.go:94]: @johnbelamaric - I'm wondering what smart thing we can do here about targeting to avoid breaking stuff.
Suggestions?

[Should we consider downgrading the latest-milestone on these maybe?]

Comment (johnbelamaric) [reply cmd/kepval/main_test.go:94]: Are we validating ever KEP on every PR? That is, we will fail for any PR now unless we do that? Or will we only fail when one of these KEPs is touched?

I think it's reasonable to change their `latest-milestone` back to 1.20 and/or remove it in order to force the PRR to be filled.

File: keps/prod-readiness/sig-api-machinery/1040.yaml
Comment (johnbelamaric) [inline keps/prod-readiness/sig-api-machinery/1040.yaml:3]: Do we even need this field? Isn't presence of an approver enough? Would this ever be set to `false`?

That is, since this file cannot be created without approval from a PRR approver, it would seem that the presence of this in the repo is sufficient.

File: pkg/kepval/prrs/validations/yaml.go
Comment (johnbelamaric) [inline pkg/kepval/prrs/validations/yaml.go:47]: I think the validation should be:

- Find the `stage` from kep.yaml
- Check for presence and validity of PRR approver for that stage, if `latest-milestone` is > 1.20

**Review by johnbelamaric — CHANGES_REQUESTED** [review]

File: keps/prod-readiness/sig-api-machinery/1040.yaml
Comment (wojtek-t) [reply keps/prod-readiness/sig-api-machinery/1040.yaml]: Fair enough - removed

File: pkg/kepval/prrs/validations/yaml.go
Comment (wojtek-t) [reply pkg/kepval/prrs/validations/yaml.go:47]: That makes sense to me, but I think this logic doesn't belong here.

This part is more of a syntax validation.
The actual semantic validation depending on the kep.yaml is happening in the test itself - let me add that logic there.

**Review by wojtek-t — COMMENTED** [review]
Comments addressed.

I also extracted https://github.com/kubernetes/enhancements/pull/2184 out of tihs PR, since these are useful changes on its own.

File: pkg/kepval/prrs/validations/yaml.go
Comment (wojtek-t) [reply pkg/kepval/prrs/validations/yaml.go:47]: Added in the test itself (main_test.go)

File: cmd/kepval/main_test.go
Comment (wojtek-t) [reply cmd/kepval/main_test.go:94]: Yes - we exercise all keps on every PR.
I guess I will try to clean them up once this PR will be more ready to go.

**Comment by johnbelamaric** [comment]
Ok, it's getting there, modulo the additional cleanup.

One thing - how do we handle it when a PRR approver leaves and is no longer in the list? `prr-approvers-emeritus`?

**Comment by wojtek-t** [comment]
> One thing - how do we handle it when a PRR approver leaves and is no longer in the list? prr-approvers-emeritus?

Yeah - that would be my thought too. And validating that the approver is in any of that list.
Simple to do and quite intuitive imho.

File: cmd/kepval/main.go
Comment (wojtek-t) [inline cmd/kepval/main.go:27]: @johnbelamaric - what do you think about this one?
Do you know if people are using this binary?

I think that changing this binary to what the test itself is doing might not be the best option (because e.g. we're hardcoing prod-readiness directory, which makes sense for test, but not necessary for the binary).
OTOH, without that, it's not a full solution.

So while things like "kepctl" are really binaries, here what we really need is a test.

WDYT?

Comment (johnbelamaric) [reply cmd/kepval/main.go:27]: yeah, I don't know what this is for

File: cmd/kepval/main_test.go
Comment (johnbelamaric) [inline cmd/kepval/main_test.go:129]: Can we point the user to the docs (KEP template? PRR policy doc?), so they have a pointer to how to fix the failure?

File: cmd/kepval/main_test.go
Comment (sftim) [inline cmd/kepval/main_test.go:93]: Just a query: Is there a way for a KEP to target v1.21 without a KEP number?

Comment (wojtek-t) [reply cmd/kepval/main_test.go:93]: Currently kep-number is not required.
But I want to make it required eventually (but we would need to migrate all keps to the new template).

File: cmd/kepval/main_test.go
Comment (wojtek-t) [reply cmd/kepval/main_test.go:129]: Sure - that makes perfect sense. Will add once your relevant PRs will merge.

File: keps/sig-cloud-provider/1959-service-lb-class-field/kep.yaml
Comment (johnbelamaric) [inline keps/sig-cloud-provider/1959-service-lb-class-field/kep.yaml:27]: @andrewsykim FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-network/1860-kube-proxy-IP-node-binding/kep.yaml
Comment (johnbelamaric) [inline keps/sig-network/1860-kube-proxy-IP-node-binding/kep.yaml:19]: @Sh4d1 FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-network/2004-topology-aware-subsetting/kep.yaml
Comment (johnbelamaric) [inline keps/sig-network/2004-topology-aware-subsetting/kep.yaml:29]: @ckyiricky  @robscott  FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-network/2079-network-policy-port-range/kep.yaml
Comment (johnbelamaric) [inline keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: @rikatz FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-node/2043-pod-resource-concrete-assigments/kep.yaml
Comment (johnbelamaric) [inline keps/sig-node/2043-pod-resource-concrete-assigments/kep.yaml:30]: @dashpole @vikaschoudhary16 @RenaudWasTaken @fromanirh @AlexeyPerevalov FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-scheduling/1923-prefer-nominated-node/kep.yaml
Comment (johnbelamaric) [inline keps/sig-scheduling/1923-prefer-nominated-node/kep.yaml:18]: @chendave FYI, PRR approval will need to be part of the PR that sets this back to 1.21.

File: keps/sig-network/2079-network-policy-port-range/kep.yaml
Comment (rikatz) [reply keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: @johnbelamaric @wojtek-t ok, will review the KEP questionaire about the PRR. PR for the feature is open https://github.com/kubernetes/kubernetes/pull/97058 

Need to submit a new PR for the KEP with the questions revisited, right?

Comment (wojtek-t) [reply keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: Correct.
Or rather - I see that you already have filled some PRR sections (didn't look deep enough if all needed for Alpha, but potentially). So basically, you just need to have that reviewed by the PRR approvers.

File: keps/sig-scheduling/1923-prefer-nominated-node/kep.yaml
Comment (chendave) [reply keps/sig-scheduling/1923-prefer-nominated-node/kep.yaml:18]: Done, thanks for the notification.

File: keps/sig-network/2079-network-policy-port-range/kep.yaml
Comment (rikatz) [reply keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: @wojtek-t completed some more questions there.

If you don't mind being the PRR approver there, I can change the kep.yaml to have you as the prr-approver :) 

Thanks!

Comment (wojtek-t) [reply keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: I'm happy to review but I'm OOO for the next 2.5 weeks - can review in the second full week of January - please assign the PR to me directly.

Comment (rikatz) [reply keps/sig-network/2079-network-policy-port-range/kep.yaml:21]: No rush :) 

I'll assign you to the PR and mark you as the PRR reviewer in kep.yaml.

Thanks and happy new year

**Comment by johnbelamaric** [comment]
/hold cancel
/approve
/lgtm

@wojtek-t when you return please add a pointer to help folks resolve the failures, as discussed

**Comment by wojtek-t** [comment]
Opened https://github.com/kubernetes/enhancements/pull/2207 for followup.

**Pull request merged**
